src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c

```diff
@@ -10,9 +10,241 @@
+    struct sockaddr_storage ip_oib_addr;
+    struct sockaddr_storage dst_addr;
+    int                     num_devices;
+    char                    addrstr[128];
+    ucc_tl_mlx5_context_t  *mlx5_ctx;
+    ucc_base_lib_t         *lib;
+    int                     i;
+    int                     user_provided_ib;
+    int                     ib_valid;
+    const char             *dst;
+
+    ctx = &(context->mcast_context);
+    memset(ctx, 0, sizeof(ucc_tl_mlx5_mcast_coll_context_t));
+    memcpy(&ctx->params, mcast_ctx_conf, sizeof(ucc_tl_mlx5_mcast_ctx_params_t));
+
+    mlx5_ctx = ucc_container_of(context, ucc_tl_mlx5_context_t, mcast);
+    lib      = mlx5_ctx->super.super.lib;
+    ctx->lib = lib;
+
+    device_list = ibv_get_device_list(&num_devices);
+    if (!device_list || !num_devices) {
+        tl_debug(lib, "no ib devices available");
+        status = UCC_ERR_NOT_SUPPORTED;
+        goto error;
+    }
+
+    if (!strcmp(mcast_ctx_conf->ib_dev_name, "")) {
+        dev          = device_list[0];
+        devname      = (char *)ibv_get_device_name(dev);
+        ctx->devname = ucc_malloc(strlen(devname)+16, "devname");
+        if (!ctx->devname) {
+            status = UCC_ERR_NO_MEMORY;
+            goto error;
+        }
+        strncpy(ctx->devname, devname, strlen(devname));
+        strncat(ctx->devname, ":1", 16);
+        user_provided_ib = 0;
+    } else {
+        ib_valid = 0;
+        /* user has provided the devname now make sure it is valid */
+        for (i = 0; device_list[i]; ++i) {
+            if (!strcmp(ibv_get_device_name(device_list[i]), mcast_ctx_conf->ib_dev_name)) {
+                ib_valid = 1;
+                break;
+            }
+        }
+        if (!ib_valid) {
+            tl_debug(lib, "ib device %s not found", mcast_ctx_conf->ib_dev_name);
+            status = UCC_ERR_NOT_FOUND;
+            goto error;
+        }
+        ctx->devname     = mcast_ctx_conf->ib_dev_name;
+        user_provided_ib = 1;
+    }
+
+    ibv_free_device_list(device_list);
+    if (UCC_OK != ucc_tl_mlx5_probe_ip_over_ib(ctx->devname, &ip_oib_addr)) {
+        tl_debug(lib, "failed to get ipoib interface for devname %s", ctx->devname);
+        status = UCC_ERR_NOT_SUPPORTED;
+        if (!user_provided_ib) {
+            ucc_free(ctx->devname);
+        }
+        goto error;
+    }
+
+    is_ipv4     = (ip_oib_addr.ss_family == AF_INET) ? 1 : 0;
+    in_src_addr = (struct sockaddr_in*)&ip_oib_addr;
+
+    dst = inet_ntop((is_ipv4) ? AF_INET : AF_INET6,
+                    &in_src_addr->sin_addr, addrstr, sizeof(addrstr) - 1);
+    if (NULL == dst) {
+        tl_error(lib, "inet_ntop failed");
+        status = UCC_ERR_NO_RESOURCE;
+        goto error;
+    }
+
+    tl_debug(ctx->lib, "devname %s, ipoib %s", ctx->devname, addrstr);
+
+    ctx->channel = rdma_create_event_channel();
+    if (!ctx->channel) {
+        tl_error(lib, "rdma_create_event_channel failed, errno %d", errno);
```

Comment:
is it really an error? should we always be able to create even channel if device is up?

Reply:
I can change it to warning then.

Reply:
in general, if the error is not critical we shouldn't print errors or warnings. It won't pass QA with such output.

Reply:
@Sergei-Lebedev Sure, I changed it to tl_debug()